### ISSUE TITLE
webpack watch mode rebuild twice on file change  [MacOS]

### ISSUE THREAD

**Issue by Bnaya** [webpack/webpack#15431]
# Bug report

Minimal repro:
https://github.com/Bnaya/webpack5-twice-compile-why

**What is the current behavior?**
Run the project https://github.com/Bnaya/webpack5-twice-compile-why on a mac machine
(yarn && yarn start)
After the initial compile, open `src/index.js` and hit save.

You will see that webpack rebuild the project twice.

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

**Other relevant information:**
webpack version:   5.69.1
webpack-cli: 4.9.2
Node.js version: (14, 16)
Operating System: macos 12.2.1

**Comment by alexander-akait** [comment]
Duplicate https://github.com/webpack/webpack/issues/15206

**Comment by Bnaya** [comment]
I don't see how it's a duplicate, HMR is not involved

**Comment by Bnaya** [comment]
Also, it does not happen on linux

**Comment by alexander-akait** [comment]
@Bnaya Because double recompiling happens only on macos, there is more information in https://github.com/webpack/webpack/issues/15206

**Comment by Bnaya** [comment]
But WHY the watcher run twice?

**Comment by alexander-akait** [comment]
Because it is bug?

**Comment by Bnaya** [comment]
#15206 have not a fitting title. HMR broken It's only a side effect of the real issue.
You should at least edit the title to indicate the root of the issue
I'n also not clear from #15206 if you are going to fix it, or change the aggregated time

**Comment by vankop** [comment]
lets keep this open for now

**Comment by alexander-akait** [comment]
@Bnaya We keep original header because most of developer will find problem due broken HMR (because it is popular side effect), is this the real thing we should be focusing on right now, is it really that important since I told it?

We still investigate in this problem

**Comment by vankop** [comment]
lets keep https://github.com/webpack/webpack/issues/15206 for hmr bug only

**Comment by alexander-akait** [comment]
@vankop I don't see any reasons keeping two issues

**Comment by alexander-akait** [comment]
@vankop But `404` happens due double recompilation (because webpack removed HMR manifests), the root of the problem is one

**Comment by vankop** [comment]
@alexander-akait we already decided to keep last (last N) hmr updates https://github.com/webpack/webpack/pull/15405

**Comment by vankop** [comment]
this works for me fine... maybe there something else that will help to reproduce?

**Comment by Bnaya** [comment]
> this works for me fine... maybe there something else that will help to reproduce?

I have 5 developers that this minimal repro did happen for them.
Intel macs 
If you want to have a debug session on zoom or something, i would happy to.
You can reach me at me@bnaya.net

**Comment by vankop** [comment]
@Bnaya could you run webpack with
```js
infrastructureLogging: { debug: true }
```
and provide output on double compilation here?

**Comment by Bnaya** [comment]
> @Bnaya could you run webpack with
> 
> ```js
> infrastructureLogging: { debug: true }
> ```
> 
> and provide output on double compilation here?


Output:
```
    [webpack-cli] File '/private/tmp/playground/wtf-cra5/lama/src/index.js' was modified
    [webpack-cli] Changed time is Thu Jul 15 54117 20:56:43 GMT+0300 (Israel Daylight Time) (timestamp is 1645617319003)
    [webpack-cli] Compiler starting... 
    [webpack-cli] Compiler is using config: '/private/tmp/playground/wtf-cra5/lama/webpack.config.js'
asset main.js 1.2 KiB [emitted] (name: main)
./src/index.js 23 bytes [built] [code generated]
webpack 5.69.1 compiled successfully in 9 ms
    [webpack-cli] Compiler finished
    [webpack-cli] Compiler is watching files for updates...
    [webpack-cli] File '/private/tmp/playground/wtf-cra5/lama/src/index.js' was modified
    [webpack-cli] Changed time is Thu Jul 15 54117 20:56:43 GMT+0300 (Israel Daylight Time) (timestamp is 1645617319003)
    [webpack-cli] Compiler starting... 
    [webpack-cli] Compiler is using config: '/private/tmp/playground/wtf-cra5/lama/webpack.config.js'
asset main.js 1.2 KiB [emitted] (name: main)
./src/index.js 23 bytes [built] [code generated]
webpack 5.69.1 compiled successfully in 10 ms
    [webpack-cli] Compiler finished
    [webpack-cli] Compiler is watching files for updates...
    ```

**Comment by vankop** [comment]
@Bnaya increasing https://webpack.js.org/configuration/watch/#watchoptionsaggregatetimeout may help (set 200, right now default is 20)

**Comment by Bnaya** [comment]
I have already applied that workaround,
But it's still a bug here
I'm sure there is a DX reason that Sorka changed it to 20

**Comment by markjm** [comment]
Perhaps worth noting that the issue is not watchpack-specific. Can simply try watching a file on mac and see that two events are triggered

```
const fs = require("fs");
const watcher = fs.watch("./src/index.js");
watcher.on("change", (type, filename) => {
    console.log(type, filename, Date.now(), fs.statSync("./src/index.js"))
});
```

It may also be worth nothing that the issue does not appear when watching the containing directory (`fs.watch("./src/index.js")` triggers multiple times while `fs.watch("./src/")` triggers only once. Node docs do indicate some differences in watching between files and directories

https://nodejs.org/api/fs.html#fs_caveats

> This feature depends on the underlying operating system providing a way to be notified of filesystem changes.

> On Linux systems, this uses [inotify(7)](https://man7.org/linux/man-pages/man7/inotify.7.html).
> On BSD systems, this uses [kqueue(2)](https://www.freebsd.org/cgi/man.cgi?query=kqueue&sektion=2).
> **On macOS, this uses [kqueue(2)](https://www.freebsd.org/cgi/man.cgi?query=kqueue&sektion=2) for files and [FSEvents](https://developer.apple.com/documentation/coreservices/file_system_events) for directories.**

Perhaps can workaround on webpack side by only watching directories on mac and filtering un-needed events, but that may be very costly

**Comment by markjm** [comment]
This can also be resolved by setting ignoreWhenEqual here. From my understanding, that change would make sense - but maybe missing some history as to why it is set to false

https://github.com/webpack/watchpack/blob/ba21615e61ed9013429d498faf1fe21d208ae6c3/lib/DirectoryWatcher.js#L438

**Comment by daniel-nagy** [comment]
Maybe this is a bug in Node? I took the example code from [kqueue](https://www.freebsd.org/cgi/man.cgi?query=kqueue) and tested it locally and it only notifies once for file change.

![2022-02-23 12 41 28](https://user-images.githubusercontent.com/1622446/155386171-229df61b-3efd-4661-bf82-c6dd9af1f3e0.gif)

**Comment by markjm** [comment]
Agreed that it seems to be a node bug - just suggesting possible fixes/mitigations we can apply at the webpack level

**Comment by daniel-nagy** [comment]
Maybe it is not a bug, I guess it depends on the intent of the original author. In this case it is certainly unexpected and undesirable. If the `watch` API is behaving differently on different platforms then from that perspective you might consider it a bug.

Perhaps in hindsight the `watch` API should not be limited to `'change'` and `'rename'` events but maybe this was done for cross platform compatibility reasons. It would be nice to get someone from the Node team to way in here.

**Comment by daniel-nagy** [comment]
> is anyone able to repro in plain terminal without having the files open in vscode? Perhaps new issue with mac build of vscode?

Just tried with vim, same issue.

![2022-02-23 18 21 48](https://user-images.githubusercontent.com/1622446/155433309-b24ad946-094e-4eee-a4dc-01fab59d60c8.gif)

**Comment by sokra** [comment]
> This can also be resolved by setting ignoreWhenEqual here. From my understanding, that change would make sense - but maybe missing some history as to why it is set to false

Some filesystems have mtime accuracy up to 2s. Ignoring changes where mtime doesn't change would lead to missing changes when they occur within the right 2s window.

**Comment by vankop** [comment]
@Bnaya what aggregate timeout was fine for you? I think we should change defaults..

**Comment by markjm** [comment]
> > This can also be resolved by setting ignoreWhenEqual here. From my understanding, that change would make sense - but maybe missing some history as to why it is set to false
> 
> Some filesystems have mtime accuracy up to 2s. Ignoring changes where mtime doesn't change would lead to missing changes when they occur within the right 2s window.

Which filesystems? Would setting this to false make sense for macOS only?

As an aside - has anyone opened a bug on node yet for this inconsistency?

**Comment by sokra** [comment]
> Which filesystems?

FAT, but ext3 has 1s, that's also not the best.
I also assume that mtime is only set once even if the file access is splitted into multiple writes. So I guess you could still see different file content for the same mtime even with high accuracy.

> It may also be worth nothing that the issue does not appear when watching the containing directory

btw. watchpack never watches files, but only watches directories.

**Comment by Bnaya** [comment]
> @Bnaya what aggregate timeout was fine for you? I think we should change defaults..

I didn't find a magic number that i feel ok with,
I went with 400, it feels too much, but better than the alternative

**Comment by kubaprzetakiewicz** [comment]
Is there any update to this?

We've set `watchOptions.aggregateTimeout` to `400` - yet we're still marred with double rebuild and HMR issues stemming from it.

<details>
<summary>
Environment details:
</summary>

```
➜ ./node_modules/.bin/webpack --version
webpack: 5.70.0
webpack-cli: 4.9.2
webpack-dev-server 4.7.4

➜ node -v
v16.13.1

➜ sw_vers
ProductName:	macOS
ProductVersion:	12.2.1
BuildVersion:	21D62
```

</details>

**Comment by markjm** [comment]
Perhaps not the perfect solution for some of those other less common (i think?) file systems, but I think the linked PR should resolve this issue while still respecting lower accuracy mtimes

**Comment by kubaprzetakiewicz** [comment]
> @kubaprzetakiewicz Could you provide reproducible repo?

https://github.com/levymetal/webpack5-double-compile-repro

you can fiddle with the `aggregateTimeout` here and it's sometimes reproducible

**Comment by alexander-akait** [comment]
We have patch https://github.com/webpack/watchpack/pull/215, somebody can try it locally?

**Comment by Magin-CC** [comment]
> dy can try it local

I use this patch in our project, it seems work well

**Comment by Bnaya** [comment]
Thank you @markjm!!
